src/H5T.c

```diff
@@ -3394,6 +3394,7 @@ H5T__create(H5T_class_t type, size_t size)
             /* Copy the default string datatype */
             if (NULL == (dt = H5T_copy(origin_dt, H5T_COPY_TRANSIENT)))
                 HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, NULL, "unable to copy");
+            dt->shared->type = type;
```

Comment:
The logic for `H5Tcreate` with `H5T_INTEGER`, `H5T_FLOAT`, `H5T_TIME` and `H5T_STRING` weren't setting the new datatype's type to the given type, like the other cases below. This meant that newly-created types in any combination of these four categories could compare equal to each other if their sizes are the same.